aptos-move/move-examples/vote_lockup/sources/vote.move

```diff
@@ -0,0 +1,389 @@
+        certificate: Object<VotingCertificate>,
+        tokens: FungibleAsset,
+    ) acquires VoteConfig, VotingCertificate {
+        let certificate_data = unchecked_mut_certificate(&certificate);
+        let amount = fungible_asset::amount(&tokens);
+        assert!(amount > 0, EINVALID_AMOUNT);
+        let old_amount = certificate_data.locked_amount;
+        let new_amount = old_amount + amount;
+        certificate_data.locked_amount = new_amount;
+        voting_token::deposit(certificate, tokens);
+
+        // Has to called for every function that modifies amount or lockup duration of any voting certificate.
+        // Always at the end of a function so we don't forget.
+        let end_epoch = certificate_data.end_epoch;
+        update_manifested_total_supply(old_amount, end_epoch, new_amount, end_epoch);
+    }
+
+    /// Withdraw the $VOTING from a voting certificate. The certificate must have expired.
+    public entry fun withdraw_entry(
+        owner: &signer,
+        certificate: Object<VotingCertificate>,
+    ) acquires VotingCertificate {
+        let assets = withdraw(owner, certificate);
+        primary_fungible_store::deposit(signer::address_of(owner), assets);
+    }
+
+    /// Withdraw the $VOTING from a voting certificate. The certificate must have expired.
+    public fun withdraw(
+        owner: &signer,
+        certificate: Object<VotingCertificate>,
+    ): FungibleAsset acquires VotingCertificate {
+        // Extract the unlocked $VOTING and burn the ve token.
+        let tokens = voting_token::withdraw(certificate, fungible_asset::balance(certificate));
+        let end_epoch = owner_only_destroy_certificate(owner, certificate);
+        // This would fail if the lockup has not expired yet.
+        assert!(end_epoch <= epoch::now(), ELOCKUP_HAS_NOT_EXPIRED);
+        tokens
+        // Withdraw doesn't need to update total voting power because this lockup should not have any effect on any
+        // epochs, including the current one, as it has already expired.
+    }
+
+    fun update_manifested_total_supply(
+        old_amount: u64,
+        old_lockup_end_epoch: u64,
+        new_amount: u64,
+        new_lockup_end_epoch: u64,
+    ) acquires VoteConfig {
+        assert!(
+            new_amount > old_amount || new_lockup_end_epoch > old_lockup_end_epoch,
+            EINVALID_LOCKUP_CHANGE,
+        );
+
+        // We only need to update the total supply starting from the current epoch since the total voting powers of
+        // past epochs are already set in stone.
+        let curr_epoch = epoch::now();
+        let total_voting_power_per_epoch = &mut unchecked_vote_config().unscaled_total_voting_power_per_epoch;
+        while (curr_epoch < new_lockup_end_epoch) {
+            // Old epoch delta can be zero if there was no previous lockup (old_amount = 0) or lockup has expired.
+            let old_epoch_delta = if (old_amount == 0 || old_lockup_end_epoch <= curr_epoch) {
```

Comment:
should we just skip over old epochs that are < now?

Reply:
not sure why we can't bump in the current epoch?

Reply:
Current epoch is already underway so it makes more sense to not allow changing the voting power then. This is up to the application though to define the rules. Let me add more comments here to clarify this

Reply:
Also specifically here, the end epoch of a lockup is the first epoch that voting tokens can be withdrawn and voting power will be 0 (aka lockup has expired).